libraries/chain/include/eosio/chain/hotstuff.hpp

```diff
@@ -12,10 +12,46 @@ namespace eosio::chain {
 
    using hs_bitset = boost::dynamic_bitset<uint32_t>;
 
+   static digest_type get_digest_to_sign(const block_id_type& block_id, uint8_t phase_counter, const fc::sha256& final_on_qc){
+      digest_type h1 = digest_type::hash( std::make_pair( block_id, phase_counter ) );
```

Comment:
Actually you don't need the make_pair constructor!

Reply:
removing the `std::pair` constructor:

```
modified   libraries/chain/include/eosio/chain/hotstuff.hpp
@@ -24,11 +24,6 @@ namespace eosio::chain {
 
    struct view_number {
       view_number() : _block_height(0), _phase_counter(0) {}
-      explicit view_number(std::pair<uint32_t, uint8_t> data) :
-         _block_height(data.first),
-         _phase_counter(data.second)
-      {
-      }
       
       explicit view_number(uint32_t block_height, uint8_t phase_counter) :
          _block_height(block_height),
@@ -82,7 +77,7 @@ namespace eosio::chain {
       uint32_t block_num() const { return block_header::num_from_id(block_id); }
       uint64_t get_key() const { return compute_height(block_header::num_from_id(block_id), phase_counter); };
 
-      view_number get_view_number() const { return view_number(std::make_pair(block_header::num_from_id(block_id), phase_counter)); };
+      view_number get_view_number() const { return view_number(block_header::num_from_id(block_id), phase_counter); };
 
    };
 
modified   libraries/hotstuff/test/hotstuff_tools.cpp
@@ -56,8 +56,8 @@ BOOST_AUTO_TEST_CASE(view_number_tests) try {
 
   //test constructor
 
-  eosio::hotstuff::view_number vn_6 = eosio::hotstuff::view_number(std::make_pair(194217068, 2));
+  eosio::hotstuff::view_number vn_6 = eosio::hotstuff::view_number(194217068, 2);
 
   BOOST_CHECK_EQUAL(vn_5 == vn_6, true);

```

Reply:
I'm confused? How do you do this without `make_pair` or some kind of tuple struct type?

Reply:
I agree that the `std::pair` constructor of `view_number` is not needed, but what does that have to do with this `get_digest_to_sign` implementation?

Reply:
@heifner I'm not following, I didn't touch `get_digest_to_sign`. 
Also the `get_view_number is useless, it just return a copy of itself. I updated my recommended change above

Reply:
This comment chain is on `get_digest_to_sign`. I think we are in agreement.

Reply:
Oh yes, my bad, sorry. For the `get_digest_to_sign`, can we only hash a `pair` and not a `tuple`? would be nicer to hash all 3 at once.

Reply:
Yes, no reason to hash them separately. Good call.

Reply:
I would do this:
```
struct hotstuff_digest_type { 
   const block_id_type& block_id; 
   uint8_t phase_counter = 0; 
   const fc::sha256& final_on_qc;

   digest_type to_digest() const { digest_type::hash( *this ); }
}; 


FC_REFLECT(eosio::chain::hotstuff_digest_type, (block_id)(phase_counter)(final_on_qc));

Reply:
The digest simplification is great but I think it's going to change all of the hotstuff unit tests since those use hardcoded hash values. If that's the case then I think we should do that in a separate issue/PR, referencing the comment with the fix: https://github.com/AntelopeIO/leap/pull/1578#discussion_r1312439427